### PR TITLE
Stats today widget

### PR DESCRIPTION
**Pull request by planarvoid**
This PR adds a Today widget that's very similar to the All-time widget. The data source is different, the behaviour should be the same. It also extract some common logic between all three widgets. The layouts are now shared and the title (which was the only thing different between the layouts) is set programmatically. I'll have another PR with the unit tests for the view model

To test:
* Add Today widget (light/dark, various sites)
* It shows the correct data in a block view
* Decrease its width
* The block view changes into list view
* Increase its width
* The block view is back

Screenshot_1560432290

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetListProvider.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.wordpress.android.R
+import org.wordpress.android.WordPress
+import org.wordpress.android.ui.stats.OldStatsActivity
+import org.wordpress.android.ui.stats.StatsTimeframe
+import org.wordpress.android.ui.stats.refresh.lists.widget.StatsWidgetConfigureViewModel.Color
+import javax.inject.Inject
+
+class TodayWidgetListProvider(val context: Context, intent: Intent) : RemoteViewsFactory {
+    @Inject lateinit var viewModel: TodayWidgetListViewModel
+    @Inject lateinit var widgetUpdated: TodayWidgetUpdater
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetListProvider.kt]: You probably mean `widgetUpdater` here :)

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetUpdater.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        if (networkAvailable && siteModel != null) {
+            if (showColumns) {
+                views.setOnClickPendingIntent(
+                        R.id.widget_content,
+                        widgetUtils.getPendingSelfIntent(context, siteModel.id, StatsTimeframe.INSIGHTS)
+                )
+                showColumns(appWidgetManager, appWidgetId, views, siteModel)
+            } else {
+                widgetUtils.showList(
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetUpdater.kt]: Do we need to `setOnClickPendingIntent` for a list view? It looks like that currently list view just opens Stats without sending any data with intent.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt [hunk 3]
```diff
@@ -5,69 +5,65 @@ import android.appwidget.AppWidgetManager
...
         val siteId = appPrefsWrapper.getAppWidgetSiteId(appWidgetId)
         val siteModel = siteStore.getSiteBySiteId(siteId)
         val networkAvailable = networkUtilsWrapper.isNetworkAvailable()
         val layout = when (colorMode) {
-            DARK -> R.layout.stats_widget_views_dark
-            LIGHT -> R.layout.stats_widget_views_light
+            DARK -> R.layout.stats_widget_list_dark
+            LIGHT -> R.layout.stats_widget_list_light
         }
         val views = RemoteViews(context.packageName, layout)
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt]: Should we use `getLayout` from `WidgetUtils` here too?

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt [hunk 4]
```diff
@@ -5,69 +5,65 @@ import android.appwidget.AppWidgetManager
...
-        imageManager.load(awt, context, ICON, siteIconUrl ?: "", FIT_START)
+        views.setTextViewText(R.id.widget_title, resourceProvider.getString(R.string.stats_views))
+        widgetUtils.setSiteIcon(siteModel, context, views, appWidgetId)
         siteModel?.let {
-            views.setOnClickPendingIntent(R.id.widget_title, getPendingSelfIntent(context, siteModel.id))
+            views.setOnClickPendingIntent(
+                    R.id.widget_title,
+                    widgetUtils.getPendingSelfIntent(context, siteModel.id, DAY)
+            )
             views.setPendingIntentTemplate(R.id.widget_content, getPendingTemplate(context))
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt]: We are calling setPendingIntentTemplate on `widget_content` in `widgetUtils.showList` below, so I wonder is it ok to remove it here?

**Review by khaykov — CHANGES_REQUESTED** [review]
Today widget works pretty well! I have a couple of comments in the code (eg. list view does not send intent). 

I think refactoring might have brought a couple of new issues (or maybe I just missed them before?):
Views Widget sometimes loads stats for one week before. And sometimes this happens:
[Image from Gyazo](https://gyazo.com/fff94aa88024806555a23aa5411ae509)

After clicking between all-time and views widget, I noticed that after some time wrong Stats screen is shown (eg. clicking on All-Time widget show's Day stats and clicking on View header shows Insights).

Sorry, I don't really have exact steps to reproduce those, but just having a couple of different widgets for same/different sites and clicking between them works for me.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetUpdater.kt (hunk 2 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/TodayWidgetUpdater.kt]: I'll check with Sylvester about what we need to do on item click in the list. It's likely we can only use the `setOnClickPendingIntent` to open the stats and we don't need individual item click 👍

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt (hunk 3 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt]: the `Views` widget doesn't show the block layout. We can create a second method (maybe `getListLayout`) which we can reuse in all the widgets

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt (hunk 4 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/widget/ViewsWidgetUpdater.kt]: good catch 👍

**Comment by planarvoid** [comment]
@khaykov thanks for the review! I've actually noticed the wrong dates before but I couldn't find a consistent way to make them show. I've changed a bit the refresh logic so it matches the `Overview` use case. Let me know if you see the issue after the changes!

**Review by khaykov — APPROVED** [review]
Thanks for the changes! Looks like the issue with wring stats range have been fixed :) I still sometimes see stats for the wrong site being opened from the Views widget title, but I can't reliably reproduce it and don't really see a reason for it, so I hope we will figure out something along the way 👍

**Pull request merged**